### PR TITLE
[SYCL] Add online compilation extension spec.

### PR DESCRIPTION
**Pull request by kbobrovs**
Formal document for https://github.com/intel/llvm/pull/2901

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+};
+-----------------
+
+=== APIs to express compilation target characteristics
+
+The desired format of the compiled code: 
+[source,c++]
+-----------------
+enum class compiled_code_format {
+  SPIRV
```
Comment (Pennycook) [inline sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:41]: Similar to above, maybe `compiled_code_format::spir_v`?

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+|Whether the target device architecture is 64-bit.
+
+|`std::string` DeviceStepping
+|Target device stepping (implementation defined).
+
+|===
+
+Online compiler construction or source compilation may be unsuccessful, in which case an instance
+of `sycl::INTEL::online_compile_error` is thrown. For example, when some of the compilation
+target specification elements are not supported by the implementaiton, or there is a syntax error
```
Comment (Pennycook) [inline sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:102]:
```suggestion
target specification elements are not supported by the implementation, or there is a syntax error
```

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+
+==== `sycl::INTEL::online_compiler` constructors.
+[cols="40,60",options="header"]
+|===
+|Constructor |Description
+
+|`online_compiler(compiled_code_format fmt = compiled_code_format::SPIRV)`
+| Constructs online compiler which can target any device and produces
+  given compiled code format. Produces device code is 64-bit. OutputFormatVersion is
```
Comment (Pennycook) [inline sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:113]:
```suggestion
  given compiled code format. Produced device code is 64-bit. OutputFormatVersion is
```

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  implementation defined. The created compiler is "optimistic" - it assumes all applicable SYCL
+  device capabilities are supported by the target device(s).
+
+|`online_compiler(
+    sycl::info::device_type dev_type,
+    device_arch arch,
+    compiled_code_format fmt = compiled_code_format::SPIRV)`
```
Comment (Pennycook) [inline sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:120]: `spir_v` if change is made above.

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+Instantiations of the compilation function:
+[source,c++]
+-----------------
+/// Compiles given OpenCL source. May throw \c online_compile_error.
+/// @param src - contents of the source
+/// @param options - compilation options (implementation defined); standard
+///   OpenCL JIT compiler options must be supported
+template <>
+template <>
+std::vector<byte> online_compiler<source_language::OpenCL_C>::compile(
```
Comment (Pennycook) [inline sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:147]: `opencl_c` if change is made above.

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+template <>
+template <>
+std::vector<byte> online_compiler<source_language::CM>::compile(
+    const std::string &src);
+
+/// Compiles given CM source.
+/// @param options - compilation options (implementation defined)
+template <>
+template <>
+std::vector<byte> online_compiler<source_language::CM>::compile(
```
Comment (Pennycook) [inline sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:160]: `cm` if change is made above.

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+__kernel void my_kernel(__global int *in, __global int *out) {
+  size_t i = get_global_id(0);
+  out[i] = in[i] + 1;
+}
+)===";
+
+using namespace sycl::INTEL;
+
+int main(int argc, char **argv) {
+  online_compiler<source_language::OpenCL_C> compiler;
```
Comment (Pennycook) [inline sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:183]: `opencl_c` if changed above.

**Review by Pennycook — CHANGES_REQUESTED** [review]

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+All online compilation API elements reside in the `sycl::INTEL` namepace.
+
+=== Source language specification
+
+Elements of the enum designate the source language:
+[source,c++]
+-----------------
+enum class source_language {
+  OpenCL_C, // OpenCL C language
+  CM        // Intel's C-for-Media language
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:31]: ok

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc (hunk 1 above)
Comment (kbobrovs) [reply sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:41]: ok

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc (hunk 4 above)
Comment (kbobrovs) [reply sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:120]: ok

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc (hunk 5 above)
Comment (kbobrovs) [reply sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:147]: ok

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc (hunk 6 above)
Comment (kbobrovs) [reply sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:160]: ok

File: sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc (hunk 7 above)
Comment (kbobrovs) [reply sycl/doc/extensions/OnlineCompilation/OnlineCompilation.asciidoc:183]: ok

**Review by Pennycook — APPROVED** [review]

**Comment by vladimirlaz** [comment]
Could you please clarify if compilation in this document means compilation only or compilation+linkage?
 1. if first how the linkage can be done? Can we link multiple modules?
 2. if second can it be state explicitly in the doc?

**Review by v-klochkov — APPROVED** [review]

**Comment by kbobrovs** [comment]
> Could you please clarify if compilation in this document means compilation only or compilation+linkage?
> 
> 1. if first how the linkage can be done? Can we link multiple modules?
> 2. if second can it be state explicitly in the doc?

It means compilation only. Doing linkage is outside of the scope of this specification. SYCL 1.2.1/2020 provides ways to further link or otherwise handle the resulting code.

**Pull request merged**